### PR TITLE
ui/cluster-ui: move timescale components to cluster-ui

### PR DESCRIPTION
**Pull request by xinhaoz**
This commit moves timescale components, namely TimeScaleDropdown
and its related components, to cluster-ui. This change is required
as this component will be required for metrics pages in serverless
and cloud dedicated.

Previously, the timescale dropdown component managed both a
time window object containing start and end dates, as well as
a time scale object defining a window size, and the window end.
In practice, the window size prop should correlate directly
with the time scale prop, and the window should and can be
derived from the timescale object. This commit removes the
management of a separate time window from the time scale, i.e.
the timescale dropdown now manages only a timescale object.

Release note: None

-----------------------
image
custom:
image

### CODE REVIEW

**Review by maryliag — APPROVED** [review]
just one small nit and :lgtm: 

> Reviewed 19 of 20 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-:-MnVvSpOBXDXzJRhGX3q:bsgchky)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @maryliag and @xinhaoz)

___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnVukx12TI0E3t4Fx_0:-MnVukx12TI0E3t4Fx_1:b-mmvkke) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e64f776baea732a5815450c78f309054c5ed937/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx#L124)):*
> ```XML
> 
> // TimeScaleDropdown is the dropdown that allows users to select the time range
> // for graphs.
> ```

nit: this needs to be update, now is no longer just graphs, you can change to `... the time range of the data being displayed`

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, and @xinhaoz)

___
*[pkg/ui/workspaces/cluster-ui/src/dropdown/dropdown.tsx, line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnWPTv88uC3YZgx66g9:-MnWPTv88uC3YZgx66gA:b-7oc5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6dd4c9068c4a88509e71d8b2907b3ea0cc91dac/pkg/ui/workspaces/cluster-ui/src/dropdown/dropdown.tsx#L116)):*
> ```XML
>   };
> 
>   renderDropdownToggleButton = (): React.ReactNode => {
> ```

Hmm, why is this render func returns `ReactNode` while other returns `ReactElement` ? 
___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnWM2oq1a8CF8Y6rybY:-MnWM2oq1a8CF8Y6rybZ:b-z5wdt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6dd4c9068c4a88509e71d8b2907b3ea0cc91dac/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx#L1)):*
> ```XML
> // Copyright 2018 The Cockroach Authors.
> ```

nit: we might need to change the copyright header 
___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx, line 22 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnWNW1EETFQpH6U4c26:-MnWNW1EETFQpH6U4c27:ble2z8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6dd4c9068c4a88509e71d8b2907b3ea0cc91dac/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx#L22)):*
> ```XML
> const cx = classNames.bind(styles);
> 
> export type RangeOption = {
> ```

seems like we are mixing `type` and `interface` in this file, what's the convention in typescript for when to use one or another? 
___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnWNfz_D0Io0Q1aeOwi:-MnWNfz_D0Io0Q1aeOwj:b-zdjyd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6dd4c9068c4a88509e71d8b2907b3ea0cc91dac/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx#L69)):*
> ```XML
>       className={cx("_time-button", isSelected ? "active" : "")}
>       onClick={_onClick}
>       ghost
> ```

ghost ?

**Review by maryliag — CHANGES_REQUESTED** [review]
talked offline and discusses some changes, so removing the approval for now and will take a look once the changes are implemented

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-:-Mna_8aP9W6oJKzhpy0F:b-iy8b8o)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, and @xinhaoz)

**Review by xinhaoz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/dropdown/dropdown.tsx, line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnWPTv88uC3YZgx66g9:-MnaPd3y9-UdsULeybyH:b-25uwr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6dd4c9068c4a88509e71d8b2907b3ea0cc91dac/pkg/ui/workspaces/cluster-ui/src/dropdown/dropdown.tsx#L116)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm, why is this render func returns `ReactNode` while other returns `ReactElement` ? 
</blockquote></details>

This function might return something that is typed as ReactNode from the props, and from what I understood, the ReactNode type, the ReactNode type _may_ be a ReactElement, but could also be other things, which is why we have to type this return as ReactNode.
___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnWM2oq1a8CF8Y6rybY:-MnaPZgQ3o-61IBrtQL8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6dd4c9068c4a88509e71d8b2907b3ea0cc91dac/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx#L1)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: we might need to change the copyright header 
</blockquote></details>

Done.
___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx, line 22 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnWNW1EETFQpH6U4c26:-MnaMRQb4egExwGJ-Pxn:b-d87ux0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6dd4c9068c4a88509e71d8b2907b3ea0cc91dac/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx#L22)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

seems like we are mixing `type` and `interface` in this file, what's the convention in typescript for when to use one or another? 
</blockquote></details>

It is mostly a personal preference at this point: https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#differences-between-type-aliases-and-interfaces.
___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnWNfz_D0Io0Q1aeOwi:-MnaM-7u9F0u0fDE8fPo:bw430ds) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6dd4c9068c4a88509e71d8b2907b3ea0cc91dac/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx#L69)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

ghost ?
</blockquote></details>

This makes the button BG transparent.
___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnVukx12TI0E3t4Fx_0:-MnaParjBlFnvsgk_3Ru:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e64f776baea732a5815450c78f309054c5ed937/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx#L124)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

nit: this needs to be update, now is no longer just graphs, you can change to `... the time range of the data being displayed`
</blockquote></details>

Done.

**Comment by jocrl** [comment]
___
*[pkg/ui/workspaces/cluster-ui/src/dropdown/dropdown.tsx, line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnWPTv88uC3YZgx66g9:-Mo-IJVp98J5oTGaYv1g:b-gaji03) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6dd4c9068c4a88509e71d8b2907b3ea0cc91dac/pkg/ui/workspaces/cluster-ui/src/dropdown/dropdown.tsx#L116)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

This function might return something that is typed as ReactNode from the props, and from what I understood, the ReactNode type, the ReactNode type _may_ be a ReactElement, but could also be other things, which is why we have to type this return as ReactNode.
</blockquote></details>

Is the problem that `ReactElement` doesn't cover the case of a string child? If so, would `ReactChild` work? `ReactNode` is quite permissive, and I came across a CC Console doc recommending not to use it https://github.com/cockroachlabs/managed-service/blob/b95b7e8f124264ff21efe917c41ac7fd564cd24a/docs/frontend-codereview-guide.md#prop-typing

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @jocrl, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MnWNfz_D0Io0Q1aeOwi:-Mo4Yfnr8eZNNFvxMRx9:b-3xibs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6dd4c9068c4a88509e71d8b2907b3ea0cc91dac/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/rangeSelect.tsx#L69)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

This makes the button BG transparent.
</blockquote></details>

TIL :+1:

**Review by maryliag — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 3 files at r2, 10 of 11 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-:-Mo9TpJwEsRhNkOVepKZ:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @xinhaoz)

**Review by Azhng — COMMENTED** [review]
> Reviewed 1 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @xinhaoz)

___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-Mo9jcBlBI-Yf00PndtM:-Mo9jcBlBI-Yf00PndtN:b-52w28i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6007a0b54704affbe8d0068da27bbf7144ead1d/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx#L49)):*
> ```XML
>     : currentWindow
>     ? moment.duration(moment.utc(currentWindow.end).diff(currentWindow.start))
>     : moment.duration(10, "minutes");
> ```

nit: this nested ternary operator is really hard to follow here, what does it do exactly ?

**Comment by xinhaoz** [comment]
___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-Mo9jcBlBI-Yf00PndtM:-MoAB2lE1i4MwB03IrO7:b-ohb5yd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6007a0b54704affbe8d0068da27bbf7144ead1d/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx#L49)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: this nested ternary operator is really hard to follow here, what does it do exactly ? 
</blockquote></details>

Ah this was leftover from when I moved the code. Updated the function so that we have to have either windowSize or currentWindow at this point, so that should simplify things.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @maryliag and @xinhaoz)

___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-Mo9jcBlBI-Yf00PndtM:-MoADKHCCNZv6co-oRXx:bao4l6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6007a0b54704affbe8d0068da27bbf7144ead1d/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx#L49)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Ah this was leftover from when I moved the code. Updated the function so that we have to have either windowSize or currentWindow at this point, so that should simplify things.
</blockquote></details>

Nice! It's a lot better now!

Just to help with my understanding, is it possible for `currentWindow.end` to be `null` here? (e.g. when we choose to show the "Past 30 Minutes" option)

**Comment by xinhaoz** [comment]
___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-Mo9jcBlBI-Yf00PndtM:-MoB4F4F9Nz7xQpuitOO:b-nesm3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6007a0b54704affbe8d0068da27bbf7144ead1d/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeScaleDropdown.tsx#L49)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Nice! It's a lot better now!

Just to help with my understanding, is it possible for `currentWindow.end` to be `null` here? (e.g. when we choose to show the "Past 30 Minutes" option)
</blockquote></details>

Nope, even when we have past x time, the TImeWindow end property will be explicitly set/updated.

**Review by jocrl — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @maryliag, and @xinhaoz)

___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeFrameControls.module.scss, line 3 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MoEXx5i88qZ3uWwItwR:-MoEXx5i88qZ3uWwItwS:b2t1g7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/f377d72e2bc8c6b8ab50397ed7892acc0c178807/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeFrameControls.module.scss#L3)):*
> ```SCSS
> @import "src/core/index.module";
>   
> .controls-content {
> ```

It looks like this was converted from stylus to sass. Just curious, did you run it through a .styl -> .sass converter online?

**Comment by xinhaoz** [comment]
___
*[pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeFrameControls.module.scss, line 3 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MoEXx5i88qZ3uWwItwR:-MoFPE6KDa-q_8plSzCh:bk9dxnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f377d72e2bc8c6b8ab50397ed7892acc0c178807/pkg/ui/workspaces/cluster-ui/src/timeScaleDropdown/timeFrameControls.module.scss#L3)):*
<details><summary><i>Previously, jocrl (Josephine) wrote…</i></summary><blockquote>

It looks like this was converted from stylus to sass. Just curious, did you run it through a .styl -> .sass converter online?
</blockquote></details>

Nope, I just did it manually, but I'll keep that in mind for the future! Wasn't too bad with multiple cursors.

**Comment by xinhaoz** [comment]
Noticed https://github.com/cockroachdb/cockroach/pull/72431 was merged yesterday, just want to get @koorosh to take a look and make sure I preserved those changes.

**Review by koorosh — COMMENTED** [review]
@xinhaoz , changes from my PR looks good, but that PR missed one important fix so there one more PR https://github.com/cockroachdb/cockroach/pull/72723 (very small one) which I hope can be included as well.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-:-MoPRoy81leyMdoeXjAf:bbv2w1l)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @jocrl, @maryliag, and @xinhaoz)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MoPSvmu0WdPdNsYZK4M:-MoPSvmu0WdPdNsYZK4N:b-yabrv7) (no related file):*
I've noticed that design for custom date range panel is changed. Is it intentionally?
Before:


Now:

 

---

**Comment by xinhaoz** [comment]
@koorosh Changes have been included, and yes, the custom design range was switched over to the one on cluster-ui intentionally.

**Review by maryliag — COMMENTED** [review]
> Reviewed 1 of 3 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72319)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Annebirzin, @Azhng, and @jocrl)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/72319#-MoPSvmu0WdPdNsYZK4M:-MoZA4nz6lwh2_6RDBk-:bb83if2) (no related file):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

I've noticed that design for custom date range panel is changed. Is it intentionally?
Before:


Now:

 
</blockquote></details>

I'm not sure if the design range should have changed. @annebirzin can you confirm which visual we should be seeing when the custom range is selected?

---

**Comment by Annebirzin** [comment]
@maryliag Yes I think we should use the new custom picker in DB console so we are consistent across DB console and CockroachCloud. The new picker has a more clear UX for selecting a 'start' and 'end' time.

A couple of nits
- the spacing of the 'start' and 'end' labels seems a bit off ([figma designs here](https://www.figma.com/file/MoSPFkEyJfGipCmQvvLg1a/21.2_SQL-obsrv_query-performance?node-id=3136%3A100830) for reference)
- the dropdown box should left-align with the input

**Comment by xinhaoz** [comment]
bors r=maryliag

**Comment by xinhaoz** [comment]
Design changes addressed!
bors r=maryliag
